Replaced Old Apache HTTP Libraries with OkHttpClient

kunall17 opened the pull request:
Espresso Test's are already included in #45 and #48!

smarx commented:
*Automated message from Dropbox CLA bot*

@kunall17, it looks like you've already signed the Dropbox CLA. Thanks!

niftynei commented on app/src/main/java/com/zulip/android/networking/HTTPRequest.java:
```diff
@@ -1,141 +1,115 @@
-        HttpDeleteWithReq(String url) {
-            super(url);
+    public Response execute(String method, String path) throws IOException {
+        if ("".equals(method)) throw new IOException(app.getString(R.string.method_null));
```
does this also check for nulls?

niftynei commented on app/src/main/java/com/zulip/android/networking/HTTPRequest.java:
```diff
@@ -1,141 +1,115 @@
-            Log.i("HTTP.response", "<empty>");
-        }
-        return responseString;
+        return encodedUrl.substring(0, encodedUrl.length() - 1);
```
there's probably a safer way to removing the last character only if it's an `&`.

niftynei commented on app/src/main/java/com/zulip/android/networking/ZulipAsyncPushTask.java:
```diff
@@ -131,12 +136,12 @@ protected void handleError(Exception e) {
-    protected void handleHTTPError(HttpResponseException e) {
-        if (e.getStatusCode() < 500 && e.getStatusCode() >= 400) {
+    protected void handleHTTPError(Exception e) {
+//        if (e.getStatusCode() < 500 && e.getStatusCode() >= 400) {
```
If this code is unused, it should be deleted. :)

niftynei commented on app/src/main/java/com/zulip/android/networking/ZulipAsyncPushTask.java:
```diff
@@ -93,11 +96,14 @@ public void setProperty(String key, String value) {
-            return request.execute(api_path[0], api_path[1]);
-        } catch (HttpResponseException e) {
-            handleHTTPError(e);
+            Response response = request.execute(api_path[0], api_path[1]);
```
is it possible that we don't pass enough arguments to this array?

niftynei commented on app/src/main/java/com/zulip/android/networking/AsyncLogin.java:
```diff
@@ -96,7 +85,7 @@ public void run() {
             });
             // supermethod invokes cancel for us
-            super.handleHTTPError(e);
+//        super.handleHTTPError(e);
```
can remove this if it's unused.

niftynei commented on app/src/main/java/com/zulip/android/networking/AsyncGetEvents.java:
```diff
@@ -107,8 +108,10 @@ public void run() {
+                    Response httpResponse =request.execute("GET",
+                            "v1/events");
+                    String json = httpResponse.body().string();
+                    JSONObject response = new JSONObject(json);
```
may need to check the response code for success earlier, since it no longer throws an error.

niftynei commented on app/src/main/java/com/zulip/android/networking/AsyncGetEvents.java:
```diff
@@ -130,9 +133,8 @@ public void run() {
-                } catch (HttpResponseException e) {
-                    if (e.getStatusCode() == 400) {
-                        String msg = e.getMessage();
+                    if (httpResponse.code() == 400) {
```
may need to check the response code for success earlier, since it no longer throws an error.

niftynei commented on app/src/main/java/com/zulip/android/networking/HTTPRequest.java:
```diff
@@ -1,141 +1,115 @@
 
-    // Java doesn't support a request body on DELETE requests, but RFC2616 does
-    // not prohibit it.
-    class HttpDeleteWithReq extends HttpPost {
```
Here the Delete code extends from POST, but in your updated version, it appears to not support this behavior?

niftynei commented on app/src/main/java/com/zulip/android/networking/ZulipAsyncPushTask.java:
```diff
@@ -93,11 +96,14 @@ public void setProperty(String key, String value) {
-        } catch (HttpResponseException e) {
-            handleHTTPError(e);
+            Response response = request.execute(api_path[0], api_path[1]);
+            return response.body().string();
```
this might be a good place to check for http error status codes being returned, and calling `handleHTTPError`.

niftynei commented:
I'm getting this error when running it:

```
06-21 17:21:08.528 E/Error   (18516): oops
06-21 17:21:08.528 E/Error   (18516): org.json.JSONException: Value <html><title>500 of type java.lang.String cannot be converted to JSONObject
06-21 17:21:08.528 E/Error   (18516): 	at org.json.JSON.typeMismatch(JSON.java:111)
06-21 17:21:08.528 E/Error   (18516): 	at org.json.JSONObject.<init>(JSONObject.java:160)
06-21 17:21:08.528 E/Error   (18516): 	at org.json.JSONObject.<init>(JSONObject.java:173)
06-21 17:21:08.528 E/Error   (18516): 	at com.zulip.android.networking.AsyncGetEvents.run(AsyncGetEvents.java:114)
```

niftynei commented:
Overall this looks good, it needs to handle HTTP errors better (which is what's causing the bug above).

From a UI perspective, it shows me a "Message Sent" toast even though the message isn't actually sent.  This might be better addressed in a separate issue tho.

niftynei commented:
Note that this is related to issue #29

kunall17 commented on app/src/main/java/com/zulip/android/networking/AsyncGetEvents.java:
```diff
@@ -155,6 +161,7 @@ public void run() {
                 } catch (JSONException e) {
                     backoff(e);
                 }
```
Without `Thread.sleep` I receive a traceback on my server 
```
AssertionError
2016-06-22 17:00:08,438 INFO     127.0.0.1       GET     429  11ms /api/v1/events (iago@zulip.com via okhttp)
2016-06-22 17:00:08,439 INFO     status=429, data={"msg":"API usage exceeded rate limit, try again in 36.4631140232 secs","result":"error"}
, uid=iago@zulip.com
2016-06-22 17:00:08,441 ERROR    Uncaught exception GET /api/v1/events?queue_id=1466627219%3A4&last_event_id=19& (127.0.0.1)
HTTPRequest(protocol='http', host='localhost:9993', method='GET', uri='/api/v1/events?queue_id=1466627219%3A4&last_event_id=19&', version='HTTP/1.0', remote_ip='127.0.0.1', body='', headers={'Host': 'localhost:9993', 'Client': 'Android', 'Accept-Encoding': 'gzip', 'X-Forwarded-Host': '192.168.100.6:9991', 'Connection': 'close', 'Authorization': 'Basic aWFnb0B6dWxpcC5jb206bVJHRjZ1Nmt4V0RBVHo2NU1VbWJXa3AxUzZrSk9mRG0=', 'User-Agent': 'okhttp/3.2.0'})
Traceback (most recent call last):
  File "/srv/zulip-venv-cache/615b1182e7545b42f90c5ecfcad3168c4ccb6246/zulip-venv/local/lib/python2.7/site-packages/tornado/web.py", line 1042, in _execute
    getattr(self, self.request.method.lower())(*args, **kwargs)
  File "/srv/zulip/zerver/management/commands/runtornado.py", line 184, in get
    self.set_status(response.status_code)
  File "/srv/zulip-venv-cache/615b1182e7545b42f90c5ecfcad3168c4ccb6246/zulip-venv/local/lib/python2.7/site-packages/tornado/web.py", line 245, in set_status
    assert status_code in httplib.responses
 
```

I guess this is because OkHttpClient is running faster than the existing Apache HTTP Clients!
We might have to increase/decrease the sleep time!

kunall17 replied:
@timabbott After changing from Apache HTTP to okHttpClient asyncGetEvents for the queue works a lot faster and hence it floods the server, so currently I gave a 1 sec pause after every heartbeat receives, will this be the right interval time or increasing this will also do fine? 
The current app seems fine though in receiving new messages!

timabbott replied:
What we do in the web UI is this sort of exponential backoff if the client receives an error from the server (and no backoff other than the fact that the client uses longpolling (aka doesn't pass `dont_block=True` and thus is rate-limited by new events showing up at the server if it receives a success response).  Here's the code for the exponential backoff size:  https://github.com/zulip/zulip/blob/master/static/js/server_events.js#L353

Note that in this case we also show a user-facing error message overlay over the website that explains the client can't connect to the server and with a "try now" button to avoid the user having to wait up to 90s in the event that they've been off the network for a while and just got back on.  

You can see how this looks by logging into a Zulip development server and then running ctrl-C on run-dev.py and watching.

(Since this is a complicated explanation, let me know if you have any questions!)

kunall17 commented:
@niftynei  Updated the PR, Rebased to master, fixed bugs in sending a private message, added Logs to get the URL and JSON, Added error handling in pushTask!

niftynei commented on app/src/main/java/com/zulip/android/networking/AsyncGetEvents.java:
```diff
@@ -107,8 +109,24 @@ public void run() {
+                    Response httpResponse = request.execute();
+
+                    String json = httpResponse.body().string();
+                    if (!httpResponse.isSuccessful()) {
```
It might be worth also checking the status code?

kunall17 replied:
Before fetching the String by `String json = httpResponse.body().string();` ??

niftynei commented on app/src/main/res/values/strings.xml:
```diff
@@ -60,7 +60,7 @@
-
-
+    <string name="method_null">Method should not be null</string>
+    <string name="method_error">Method Not Supported</string>
```
Can you make this lower case? like `Message not supported` ?

niftynei commented on app/src/main/java/com/zulip/android/networking/HTTPRequest.java:
```diff
@@ -1,141 +1,129 @@
+            case "POST":
+                FormBody.Builder formBody = new FormBody.Builder();
+                for (Map.Entry<String, String> map : properties.entrySet()) {
+                    formBody.addEncoded(map.getKey(), map.getValue());
```
I'm running into a problem when logging in with an email.  This API `addEncoded` is transforming the `@` into the encoded version `%40`, making it such that I can't log in. :(

kunall17 replied:
Could not reproduce this problem, I am able to successfully login with the existing code (tried with zulip.tabbott.net as well)
But yeah these strings should not be encoded so I updated the code but it's weird why was it working with my build with the encodedPaths and it works in both the ways, without and with encoding!

niftynei commented:
Cool. It's working a lot better now 👍 .

I encountered two problems tho:
1) Logging in with an incorrect username & password doesn't show an error message. 😞   
2) commented in-line as well, but the form encoding is making it such that emails aren't being passed up to the server correctly.  maybe we can turn URI encoding off?

kunall17 commented:
> 1) Logging in with an incorrect username & password doesn't show an error message. :disappointed:

I had to discuss on this one, In this PR if the JSON is recieved and is not successful (status code not lies in 200) then it goes to onTaskFailed in the callback, I think this is more better as the task is failed (task failed for example incorrect login credentials) and previously it went to the onTaskCompleted even if the JSON had a status code on non 200's, which way you think is more liable to be?

This line creates the difference https://github.com/kunall17/zulip-android/blob/e50487c6a94ba8586f9d9181047af2e36e63a803/app/src/main/java/com/zulip/android/networking/ZulipAsyncPushTask.java#L109

niftynei commented:
@kunall17 Failing the task sounds reasonable to me, as long as you can preserve the error messages appropriately.  I checked this out again though and it still doesn't show an error message with incorrect credentials.  Is this supposed to be fixed?

Managed to log in with out a problem this time, thanks for fixing that encoding error! 👍 

If you can fix up the error messages & rebase this to master (the DevBackends code is still using the old style HTTP stuff), I'll get this merged. Thanks!

kunall17 commented:
@niftynei yeah I left it purposely here, I will complete this and rebase it!

kunall17 commented:
@niftynei updated the code!
